torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -776,282 +768,151 @@ uint64_t ProcessGroupNCCL::getSequenceNumberForGroup() {
-
-      for (auto it = workMetaList_.begin(); it != workMetaList_.end();
-           /* no increment*/) {
-        auto& work = *it;
-
-        if (desyncDebug_ && !work.exception()) {
-          if (!work.startTraceUpdated_ && work.isStarted() &&
-              !terminateProcessGroup_.load() && !storeError_) {
-            work.startTraceUpdated_ = true;
-            storeError_ = !c10d::traceUpdate(
-                store_,
-                traceKeyStart_,
-                work.seq_,
-                opTypeToString(work.opType_));
-          }
+    for (auto it = workMetaList_.begin(); it != workMetaList_.end();
+         /* no increment*/) {
+      auto& work = *it;
+      work.checkAndSetException();
+      bool timedOut = work.checkTimeout();
+
+      // If work hits an exception (either an error or timeout)
+      if (work.exception()) {
+        // Abort work and corresponding communicators
+        work.abort();
+        // PG level abort, which would abort all other communicators on this
+        // rank
+        abort();
+        // Report desync state in case of timeout
+        if (desyncDebug_ && timedOut) {
+          auto desyncMsg = retrieveDesyncReport(store_, "NCCL", rank_, size_);
+          LOG(ERROR) << desyncMsg;
         }
+        // Throw exception
+        work.handleException(asyncErrorHandling_);
+      }
 
+      // Work status logging for desync debug
+      if (desyncDebug_) {
+        if (work.isStarted()) {
+          logWorkStart(work);
+        }
         if (work.isCompleted()) {
-          if (desyncDebug_ && !work.exception()) {
-            // To close the window between the check of work.isStarted() and
-            // the check of work.isCompleted().
-            if (!work.startTraceUpdated_ && !terminateProcessGroup_.load() &&
-                !storeError_) {
-              storeError_ = !c10d::traceUpdate(
-                  store_,
-                  traceKeyStart_,
-                  work.seq_,
-                  opTypeToString(work.opType_));
-            }
-            if (!terminateProcessGroup_.load() && !storeError_) {
-              storeError_ = !c10d::traceUpdate(
-                  store_,
-                  traceKeyEnd_,
-                  work.seq_,
-                  opTypeToString(work.opType_));
-            }
-          }
-          // Handle Exceptions on failed GPU operations and remove completed
-          // workNCCL objects from work vector.
-          if (!terminateProcessGroup_.load()) {
-            work.handleNCCLGuard(asyncErrorHandling_);
-          }
-          doneWorks.push_back(std::move(*it));
-          it = workMetaList_.erase(it);
-        } else {
-          // Increment the iterator if the current WorkNCCL object is not
-          // completed.
-          ++it;
+          logWorkEnd(work);
         }
       }
-      done = workMetaList_.empty();
+
+      // Clean up completed work
+      if (work.isCompleted()) {
+        it = workMetaList_.erase(it);
```

Comment:
The reason we had two separate threads was that this line would invoke the destructor of WorkNCCL resulting in destroying CUDAEvents here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp#L165. This would in turn invoke `cudaEventDestroy` here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/CUDAEvent.h#L53.

When there is a failure, NCCL kernels would block the GPU and the `cudaEventDestroy` call would actually get stuck on the host side resulting in this thread being stuck at this line. This means we would never recover from the failure since this thread never gets an opportunity to call `ncclCommAbort`

Reply:
Thanks for the comment!
I think there are two things worth considering:
1. CUDA is evolving and its APIs expose less synchronous behaviors today. For example, for the `cudaEventDestroy` API you mentioned, it should be asynchronous today.
2. My general feeling is that the addition of threads does not guarantee the whole thing to be less prone to CUDA synchronization, because the added thread could be also hanging somewhere due to other CUDA calls. It is hard to ask all contributors to together make sure certain thread is “CUDA call free.” Maybe after years of OSS development that is already no longer true. So it might not be too bad an idea to trade that for simplicity plus straightforward behavior, rather than relying on bouncing back and forth between two threads.
Just my two cents :)

Reply:
I do agree with 1. and if we can do that it would be great. But even if we do that there could be other CUDA calls that might create issues.

I do believe the current behavior with a single thread is strictly worse since it is very likely this thread would get stuck at this line and never recover from the situation. In fact we saw this exact issue in several training jobs where a single thread like this would get stuck often. That was actually the motivation to have a separate "CUDA free" thread whose responsibility was to only call `ncclCommAbort`.

Maybe we can put as much logic as we want in this thread and then have a very simple separate thread that is responsible for calling `ncclCommAbort` and have clear documentation as to why we need that thread.

Reply:
> Also, note that when NCCL kernels get stuck even async CUDA methods get stuck on the host side. 

I am not sure how likely that is today. Even if it happens, it may only happen in extreme cases. If we hit it in non-extreme cases, we should ask CUDA to fix it. After all, NCCL kernels are just like any CUDA kernels. 

> For example, cudaLaunchKernel is supposed to be async but it gets stuck on the host side when NCCL kernels are stuck.

That's less likely now. Plus, this thread is not calling `cudaLaunchKernel`.

My general take is that we should not need to maintain a complex programming model just for covering up theoretical caveats, esp when the library has evolved and continues such. On the other hand, there could be deadlocking risk coming from the thread locks used between the watchdog thread and the work cleanup thread.

Regarding keeping the watchdog thread "CUDA call free", I still don't see a sustainable way. In fact, the first thing the watchdog needs to do is to call `cudaStreamQuery` to see if the collective has completed or timed out.

Reply:
> I am not sure how likely that is today. Even if it happens, it may only happen in extreme cases. If we hit it in non-extreme cases, we should ask CUDA to fix it. After all, NCCL kernels are just like any CUDA kernels.

Was curious if we have tested how likely this is today? This used to happen very often when NCCL kernels would get stuck.

> Regarding keeping the watchdog thread "CUDA call free", I still don't see a sustainable way. In fact, the first thing the watchdog needs to do is to call cudaStreamQuery to see if the collective has completed or timed out.

We don't need this in the watchdog thread though right? All the watchdog thread has to call is `ncclCommGetAsyncError` and then `ncclCommAbort` if there are errors.